internal_packages/keybase/lib/pgp-key-store.cjsx

```diff
@@ -0,0 +1,391 @@
+      msg = _.findWhere(@_msgCache, {clientId: message.clientId})
+      if msg.timeout > Date.now()
+        return msg.body
+
+    # otherwise
+    return null
+
+  hasEncryptedComponent: (message) ->
+    if not message.body?
+      return false
+
+    # find a PGP block
+    pgpStart = "-----BEGIN PGP MESSAGE-----"
+    pgpEnd = "-----END PGP MESSAGE-----"
+
+    blockStart = message.body.indexOf(pgpStart)
+    blockEnd = message.body.indexOf(pgpEnd)
+    # if they're both present, assume an encrypted block
+    return (blockStart >= 0 and blockEnd >= 0)
+
+  decrypt: (message) =>
+    # decrypt a message, cache the result
+    # (asynchronous)
+
+    # check to make sure we haven't already decrypted and cached the message
+    # note: could be a race condition here causing us to decrypt multiple times
+    # (not that that's a big deal other than minor resource wastage)
+    if @getDecrypted(message)?
+      return
+
+    if not @hasEncryptedComponent(message)
+      return
+
+    # fill our keyring with all possible private keys
+    ring = new pgp.keyring.KeyRing
+    # (the unbox function will use the right one)
+
+    for key in _.pluck(@privKeys(), "key")
+      ring.add_key_manager(key)
+
+    # find a PGP block
+    pgpStart = "-----BEGIN PGP MESSAGE-----"
+    blockStart = message.body.indexOf(pgpStart)
+
+    pgpEnd = "-----END PGP MESSAGE-----"
+    blockEnd = message.body.indexOf(pgpEnd) + pgpEnd.length
+
+    pgpMsg = message.body.slice(blockStart, blockEnd)
+
+    # There seemed to be a problem where '+' was being encoded, and there are
+    # some potential issues with HTML tags being added to the message
+    # pgpMsg = pgpMsg.replace(/&#43;/gm,'+')
+    # pgpMsg = pgpMsg.replace(/<[^>]*>/gm,'')
+    # however, it appears the problem has disappeared
+
+    pgp.unbox { keyfetch: ring, armored: pgpMsg }, (err, literals, warnings, subkey) =>
+      if err
+        console.warn err
+        @_msgStatus.push({"clientId": message.clientId, "time": Date.now(), "message": "Unable to decrypt message."})
+      else
+        if warnings._w.length > 0
+          console.warn warnings._w
+
+        if literals.length > 0
+          plaintext = literals[0].toString('utf8')
+          pre = """
+                <style>
+                  div.decrypted {
+                    border: 3px solid rgb(121, 212, 91);
+                    border-radius: 4px;
+                    padding: 8px 12px;
+                    box-sizing: border-box;
+                  }
+                </style>
+                <div class="decrypted">
+                """
+
+          post = """
+                 </div>
+                 """
+          # can't use _.template :(
+          body = message.body.slice(0, blockStart) + pre + plaintext + post + message.body.slice(blockEnd)
+
+          # TODO if message is already in the cache, consider updating its TTL
+          timeout = 1000 * 60 * 30 # 30 minutes in ms
```

Comment:
The timeout is to ensure that decrypted messages don't hang around forever - so that you don't have to close and re-open N1 to re-encrypt your messages and ensure they're "safe" (ish, the decrypted version is still in memory, although we should perhaps change that). Same goes for private keys.